polar-core/src/polar.rs

```diff
@@ -255,42 +253,23 @@ impl Polar {
                 .collect(),
         );
 
-        // Attach context to ResourceBlock, SingletonVariable, and UnregisteredClass errors.
-        //
-        // TODO(gj): can we attach context to *all* errors here since all errors will be parse-time
-        // errors and so will have some source context to attach?
-        for diagnostic in &mut diagnostics {
-            if let Diagnostic::Error(e) = diagnostic {
-                use {ErrorKind::Validation, ValidationError::*};
-                match e.kind {
-                    Validation(ResourceBlock { ref term, .. })
-                    | Validation(SingletonVariable { ref term, .. })
-                    | Validation(UnregisteredClass { ref term, .. }) => {
-                        *e = kb.set_error_context(term, e.clone());
-                    }
-                    _ => (),
-                }
-            }
-        }
-
-        // TODO(gj): the below check is actually too eager to bomb out now -- I think we only need
-        // to bomb out if we encountered a ParseError. ValidationErrors should be fine to continue.
-
-        // Generate appropriate rule_type definitions using the types contained in policy resource
-        // blocks.
-        kb.create_resource_specific_rule_types();
-
         // TODO(gj): need to bomb out before rule type validation in case additional rule types
         // were defined later on in the file that encountered the `ParseError`. Those additional
         // rule types might extend the valid shapes for a rule type defined in a different,
         // well-parsed file that also contains rules that don't conform to the shapes laid out in
         // the well-parsed file but *would have* conformed to the shapes laid out in the file that
         // failed to parse.
-        if diagnostics.iter().any(Diagnostic::is_error) {
+        if diagnostics.iter().any(Diagnostic::is_parse_error) {
+            // NOTE(gj): need to set context _before_ clearing the KB so we still have source info.
```

Comment:
Actually, why do we need to clear the KB? What can someone do with a partially-loaded KB?

Reply:
Probably don't _need_ to, but it feels slightly more coherent to end up with an empty KB in the presence of unrecoverable errors instead of ending up with some indeterminate subset of your policy loaded.

Reply:
Yeah, okay that's probably fine.